cmd/internal/backup/providers/gcp/gcp.go

```diff
@@ -147,63 +153,51 @@ func (b *BackupProviderGCP) CleanupBackups(_ context.Context) error {
 }
 
 // DownloadBackup downloads the given backup version to the specified folder
-func (b *BackupProviderGCP) DownloadBackup(ctx context.Context, version *providers.BackupVersion, outDir string) (string, error) {
+func (b *BackupProviderGCP) DownloadBackup(ctx context.Context, version *providers.BackupVersion, writer io.Writer) error {
 	gen, err := strconv.ParseInt(version.Version, 10, 64)
 	if err != nil {
-		return "", err
+		return err
 	}
 
 	bucket := b.c.Bucket(b.config.BucketName)
 
-	downloadFileName := version.Name
-	if strings.Contains(downloadFileName, "/") {
-		downloadFileName = filepath.Base(downloadFileName)
-	}
-
-	backupFilePath := filepath.Join(outDir, downloadFileName)
-
-	b.log.Info("downloading", "object", version.Name, "gen", gen, "to", backupFilePath)
+	b.log.Info("downloading", "object", version.Name, "gen", gen)
 
 	r, err := bucket.Object(version.Name).Generation(gen).NewReader(ctx)
 	if err != nil {
-		return "", fmt.Errorf("backup not found: %w", err)
+		return fmt.Errorf("backup not found: %w", err)
 	}
 	defer r.Close()
 
-	f, err := b.fs.Create(backupFilePath)
-	if err != nil {
-		return "", err
-	}
-	defer f.Close()
-
-	_, err = io.Copy(f, r)
+	_, err = io.Copy(writer, r)
 	if err != nil {
-		return "", fmt.Errorf("error writing file from gcp to filesystem: %w", err)
+		return fmt.Errorf("error writing file from gcp to filesystem: %w", err)
 	}
 
-	return backupFilePath, nil
+	return nil
 }
 
 // UploadBackup uploads a backup to the backup provider
-func (b *BackupProviderGCP) UploadBackup(ctx context.Context, sourcePath string) error {
+func (b *BackupProviderGCP) UploadBackup(ctx context.Context, reader io.Reader) error {
 	bucket := b.c.Bucket(b.config.BucketName)
 
-	r, err := b.fs.Open(sourcePath)
-	if err != nil {
-		return err
+	destination := ProviderConstant
+	if b.compressor != nil {
+		destination += b.compressor.Extension()
+	}
+	if b.encrypter != nil {
+		destination += b.encrypter.Extension()
 	}
```

Comment:
Maybe it makes sense to have a function for this and then only store the backup object name in the struct of a provider, such that `compressor` and `encrypter` are not available in the functions of a provider implementation.